File: semgrep_asvs/rules/custom/go/constant-time-compare.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
)

func verify(k, m, sig []byte, hexsig string) bool {
	mac := hmac.New(sha256.New, k)
	mac.Write(m)
	// ruleid: constant-time-compare
	if bytes.Equal(mac.Sum(nil), sig) {
		return true
	}
	// ruleid: constant-time-compare
	if hex.EncodeToString(mac.Sum(nil)) == hexsig {
		return true
	}
	// ok: constant-time-compare
	return hmac.Equal(mac.Sum(nil), sig)
}

File: semgrep_asvs/rules/custom/php/constant-time-compare.php
<?php
// ruleid: constant-time-compare
if (hash_hmac('sha256', $m, $k) === $sig) {}
// ruleid: constant-time-compare
if ($sig != hash_hmac('sha256', $m, $k)) {}
// ok: constant-time-compare
if (hash_equals(hash_hmac('sha256', $m, $k), $sig)) {}
